frontend/catalyst/cuda/catalyst_to_cuda_interpreter.py

```diff
@@ -0,0 +1,851 @@
+    from_catalyst_to_cuda = {
+        "CRX": "crx",
+        "CRY": "cry",
+        "CRZ": "crz",
+        "PauliX": "x",
+        "PauliY": "y",
+        "PauliZ": "z",
+        "Hadamard": "h",
+        "S": "s",
+        "T": "t",
+        "RX": "rx",
+        "RY": "ry",
+        "RZ": "rz",
+        "SWAP": "swap",
+        # "CSWAP": "cswap", Bug in CUDA quantum. CSWAP is not exposed.
+        # Other instructions that are missing:
+        # ch
+        # sdg
+        # tdg
+        # cs
+        # ct
+        # r1
+    }
+
+    # From the definition of qinst_p
+    # Operands:
+    # * qubits_or_params
+    invals = safe_map(ctx.read, eqn.invars)
+    qubits_or_params = invals
+
+    # And two parameters:
+    # * op=None
+    # * qubits_len=-1
+    params = eqn.params
+    op = params["op"]
+    cuda_inst_name = from_catalyst_to_cuda[op]
+    qubits_len = params["qubits_len"]
+
+    # Now, we can map to the correct op
+    # For now just assume rx
+    cuda_inst(ctx.kernel, *qubits_or_params, inst=cuda_inst_name, qubits_len=qubits_len)
+
+    # Finally determine how many are qubits.
+    qubits = qubits_or_params[:qubits_len]
+
+    # Now that we did this, we need to remember
+    # that handle_rx is not in SSA and will not return qubits
+    # And so the eqn.outvars should be replaced with something.
+    # Let's just replace them with the input values.
+    safe_map(ctx.write, eqn.outvars, qubits)
+
+
+def change_compbasis(ctx, eqn):
+    """Compbasis in Catalyst essentially is the default observable."""
+    assert eqn.primitive == compbasis_p
+
+    # From compbasis_p's definition, its operands are:
+    # * qubits
+    qubits = safe_map(ctx.read, eqn.invars)
+
+    # We dont have a use for compbasis yet.
+    # So, the evaluation of it might as well just be the same.
+    outvals = [AbstractObs(len(qubits), compbasis_p)]
+    safe_map(ctx.write, eqn.outvars, outvals)
+
+
+def change_get_state(ctx, eqn):
+    """Change Catalyst's state_p to CUDA-quantum's state primitive."""
+    assert eqn.primitive == state_p
+
+    # From state_p's definition, its operands are:
+    # * an observable
+    # * a shape
+    invals = safe_map(ctx.read, eqn.invars)
+    # Just as state_p, we will only support compbasis.
+    obs_catalyst = invals[0]
+    # This is an assert as opposed to raising an error,
+    # because there appears to be no code that can lead to this
+    # outcome.
+    assert obs_catalyst.primitive == compbasis_p
+
+    # We don't really care too much about the shape
+    # It is only used for an assertion in Catalyst.
+    # So, we will ignore it here.
+
+    # To get a state in cuda we need a kernel
+    # which does not flow from eqn.invars
+    # so we get it from the parameter.
+    cuda_state = cudaq_getstate(ctx.kernel)
+    outvals = [cuda_state]
+    outvariables = [ctx.new_variable(AbsCudaQState())]
+    safe_map(ctx.replace, eqn.outvars, outvariables)
+    safe_map(ctx.write, eqn.outvars, outvals)
```

Comment:
Can we somehow automate these "safe" mappings (required in almost all of `change_*` methods) and remove these explicit calls?

Reply:
I will be changing the call from `safe_map` to `map`. [`safe_map`](https://github.com/google/jax/blob/07d793d7b26f5973621a2eb253b2511f15530422/jax/_src/util.py#L83-L88) is just a wrapper around map that guarantees that parameters 2 and 3 have the same length. I am changing it because `safe_map` is part of the unstable API [(although `safe_map` also has a C++ faster implementation and may be stable enough)](https://github.com/google/jax/commit/0dbd467ceaf780ed70f749d935df163dd94ce5dc)

And I am not sure what you mean by automate. Do you mean you don't want the explicit calls to `map` here?

Reply:
Removed safe_map https://github.com/PennyLaneAI/catalyst/pull/477/commits/9d84eacb81f701e242a188a40e596c648f979eee